algebra/src/bytes.rs

```diff
@@ -255,16 +287,52 @@ impl FromBytes for bool {
     }
 }
 
+impl<T: ToBytes> ToBytes for Option<T> {
+    #[inline]
+    fn write<W: Write>(&self, mut writer: W) -> IoResult<()> {
+        self.is_some().write(&mut writer)?;
+        if self.is_some() {
+            self.as_ref().unwrap().write(&mut writer)?;
+        }
+        Ok(())
+    }
+}
+
+impl<T: FromBytes> FromBytes for Option<T> {
+    #[inline]
+    fn read<R: Read>(mut reader: R) -> IoResult<Option<T>> {
+        let exists = bool::read(&mut reader)?;
+        Ok(if exists {
+            Some(T::read(&mut reader)?)
+        } else {
+            None
+        })
+    }
+}
+
 impl<T: ToBytes> ToBytes for Vec<T> {
     #[inline]
     fn write<W: Write>(&self, mut writer: W) -> IoResult<()> {
+        (self.len() as u64).write(&mut writer)?;
         for item in self {
             item.write(&mut writer)?;
         }
         Ok(())
     }
 }
 
+impl<T: FromBytes> FromBytes for Vec<T> {
+    #[inline]
+    fn read<R: Read>(mut reader: R) -> IoResult<Vec<T>> {
+        let mut res = vec![];
+        let count = u64::read(&mut reader)?;
```

Comment:
What if the Vec is bigger than 2^64 ?

Reply:
Mmm, vector bigger than 2^64 is 16777216 TB of data, do we really need to take care about such situation?

Reply:
So it will always be smaller, then you are using an oversized u64. I would remove the serialization/deserialization for Vec<T> because it highly depends on the use case so we should let the user taking care of it

Reply:
This implementation is removed